### PR TITLE
Free functions: Add syntax/semantic tests to document behavior of red…

### LINKED ISSUE #9840: [Testing] Document behavior of free functions that implicitly "override" imported namesake free functions
## Description

TBD.

### PR DESCRIPTION
**Pull request by bshastry**
…efined functions in single or across multiple source units

Fixes #9840

### CODE REVIEW

File: test/libsolidity/syntaxTests/freeFunctions/free_identical.sol [hunk 1]
```diff
@@ -0,0 +1,2 @@
+function f() pure returns (uint) { return 1337; }
+function f() pure returns (uint) { return 42; }
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/freeFunctions/free_identical.sol:2]: This should definitely not be allowed - maybe overloading, but definitely not a plain redefinition...
I'm rather surprised that this didn't have tests...

File: test/libsolidity/syntaxTests/imports/name_clash_imported_free_function.sol [hunk 2]
```diff
@@ -0,0 +1,10 @@
+==== Source: s1.sol ====
+function f() pure returns (uint) { return 1337; }
+==== Source: s2.sol ====
+import "s1.sol";
+function f() pure returns (uint) { return 1; }
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/imports/name_clash_imported_free_function.sol:5]: This should also definitely not be allowed.
It's not allowed to have ``contract C {}`` in ``s1`` and ``s2`` and this should just behave the same...
Again overloading may be an option, not sure if that was the planned or not or is wise or not, but definitely not identical functions...

**Review by ekpyron — COMMENTED** [review]
@chriseth @axic Am I missing something here or are all of these rather severe bugs we should fix as soon as possible?

**Comment by ekpyron** [comment]
I'd say this might even be worth a buglist entry...

File: test/libsolidity/semanticTests/freeFunctions/redefinition_mutability.sol [hunk 3]
```diff
@@ -0,0 +1,11 @@
+function f() pure returns (uint) { return 1337; }
+function f() view returns (uint) { return 42; }
```
Comment (chriseth) [inline test/libsolidity/semanticTests/freeFunctions/redefinition_mutability.sol:2]: Should be a declaration error - we do not do redefinitions.

File: test/libsolidity/semanticTests/multiSource/free_function_resolution_base_contract.sol [hunk 4]
```diff
@@ -0,0 +1,19 @@
+==== Source: s1.sol ====
+function f() pure returns (uint) { return 1337; }
+contract C {
+  function g() public pure returns (uint) {
+    return f();
+  }
+}
+==== Source: s2.sol ====
+import "s1.sol";
+function f() pure returns (uint) { return 1; }
```
Comment (chriseth) [inline test/libsolidity/semanticTests/multiSource/free_function_resolution_base_contract.sol:10]: Same here: declaration error.

File: test/libsolidity/semanticTests/multiSource/free_function_resolution_override_virtual.sol [hunk 5]
```diff
@@ -0,0 +1,19 @@
+==== Source: s1.sol ====
+function f() pure returns (uint) { return 1337; }
+contract C {
+  function g() public pure virtual returns (uint) {
+    return f();
+  }
+}
+==== Source: s2.sol ====
+import "s1.sol";
+function f() pure returns (uint) { return 1; }
```
Comment (chriseth) [inline test/libsolidity/semanticTests/multiSource/free_function_resolution_override_virtual.sol:10]: declaration error

**Comment by chriseth** [comment]
We should test that just importing the contract works and that the correct free function is called with regards to super and so on.

**Comment by leonardoalt** [comment]
Is this going to be merged before the fix?

**Comment by chriseth** [comment]
I  would say this is obsolete - @bshastry?

**Comment by bshastry** [comment]
I agree this is obsolete. Now that #9851 fixes the underlying issue

**Pull request closed without merging**